### PR TITLE
format_pattern and other api alignment work, fixes #108

### PR DESCRIPTION
**Pull request by Pike**

### CODE REVIEW

**Comment by spookylukey** [comment]
@Pike - some early comments.

I followed the `format_pattern` discussion at a distance, and I can see that some client code benefits from `get_message` followed by `format_pattern`, but what is the motivation behind removing `format("message", args)` and `format("message.attribute", args)`? The replacements are pretty horrendous e.g. from the patch for the tests:

```
-        val, errs = self.ctx.format('from-arg', {'arg': 123456.0})
+        val, errs = self.bundle.format_pattern(self.bundle.get_message('from-arg').value, {'arg': 123456.0})
```
The API for the normal case is now, schematically,

```
    obj.method(obj.other_method(arg).property, more_args)
```
instead of:
```
    obj.method('arg', more_args)
```

That is not a user oriented API design! Why would we do this to our users? 

It's particularly bad as I can't see the `format_pattern` and `get_message` APIs being wanted by the vast majority of users. The 'push' pattern has particular relevance for fluent-dom, but probably not to many other use cases. Does the Python API need to be dominated by a use case that is very unlikely for Python users, just for parity with Javascript? By all means add the new methods, but why remove the old? I for one would be extremely embarrassed trying to write release notes for `fluent.runtime` that attempted to sell a backwards incompatible breaking change like as if it were an improvement.

**Comment by Pike** [comment]
I agree that there's no cookies in sight right now. But I'm baking some right now.

As for API surface, we should establish a an API surface for developers aside from `FluentBundle`, and there should be something akin to `Localization` in `@fluent/dom` or `fluent-fallback` in js and rust, resp. Implementing these might be easier with just a base API on bundle. But the actual cookie I'm baking right now is `format_to_parts`.

That one should get a healthy amount of logic outside of the resolver into the bindings code, I hope.

But yes, I know that I need code to show those promises, which is why I've put this up as WIP only.

**Comment by Pike** [comment]
Thinking about this some more, one thing we've done on the JS side was to land the new API on a branch. Because, yes, we weren't sure it'd work out great for Rust, and we didn't want JS to ruin Rust's API needs.

So we decided to work on a branch, and once that branch was in a good state, we decided on whether to merge that or not.

The reviews towards that branch were "technically correct, if we go with this API, this can land as is".

Would that help here, too?

And to be explicit, that has an option to not land that branch, and to decide that the python ecosystem needs an API that diverges from that of Rust and JavaScript.

**Comment by spookylukey** [comment]
OK, that sounds like a good plan, I'll reserve judgement until I see the bigger picture.

Also, I have in my mind the idea that my compiler branch might actually want to diverge in terms of API from the resolver. There are now several accumulating things that are making the resolver and the compiler uncomfortable with each other in terms of API.

1. For the resolver `get_message` returns an AST object that is used by `format_pattern`. For the compiler, neither of these really makes sense - we don't use an AST object as part of the `format` mechanism.

2. The compiler has to keep sources (or AST objects) around indefinitely in case someone does an `add_messages` call, in which case we have to re-compile everything. For long running processes, this is unnecessary memory usage. 

3. We also need a run-time check for freshness (do we need to re-compile or not?), which for simple messages is a very measurable overhead (about 10%). I've actually eliminated this in my branch, at the cost of some [slightly tricky to follow code](https://github.com/django-ftl/python-fluent/blob/django-ftl-0.11/fluent.runtime/fluent/runtime/__init__.py#L140). I'd rather not have to do this.

What the compiler really wants to do is have a constructor that takes all the FTL source in one go, compiles it, then throws away the sources and AST to save memory. It then wants to expose something like the current `format` API. (As an optimization for the really hot inner loops, it might want an API that directly returns the function that a message has been compiled to).

**Comment by Pike** [comment]
I've rebased this and created the `next-gen-api` branch for this work.

I'm a little torn on using a distinct `_compiled` bundle member. I think there's nothing bad in using the `messages` for both AST and Resolver objects, even in multi-threaded environments.

Luke, we're starting to get real about using Fluent on bedrock, aka, www.mozilla.org. I'm going to have a tight loop with @pmac, who's leading the integration work there.

The process we're using to migrate bedrock will keep the existing l10n infra in place while we're adding Fluent to parts of existing pages. Plus, the site is using Jinja templates, so we're not going to use django-ftl. We might also end up with some different design decisions in the integration, which might be interesting to review and compare.

**Comment by Pike** [comment]
> OK, that sounds like a good plan, I'll reserve judgement until I see the bigger picture.
> 
> Also, I have in my mind the idea that my compiler branch might actually want to diverge in terms of API from the resolver. There are now several accumulating things that are making the resolver and the compiler uncomfortable with each other in terms of API.
> 
>     1. For the resolver `get_message` returns an AST object that is used by `format_pattern`. For the compiler, neither of these really makes sense - we don't use an AST object as part of the `format` mechanism.

It returns a `fluent.runtime.resolver.Message` object, and it's `value` is a subclass of `fluent.runtime.resolver.BaseResolver`.

>     2. The compiler has to keep sources (or AST objects) around indefinitely in case someone does an `add_messages` call, in which case we have to re-compile everything. For long running processes, this is unnecessary memory usage.

This comment made me think about folding `_compiled` into `messages`. We don't need to keep the AST around in the resolver, as we only transform messages and terms individually. We'd have to do an `isinstance(entry, BaseResolver)` check in `_lookup` then.

>     3. We also need a run-time check for freshness (do we need to re-compile or not?), which for simple messages is a very measurable overhead (about 10%). I've actually eliminated this in my branch, at the cost of some [slightly tricky to follow code](https://github.com/django-ftl/python-fluent/blob/django-ftl-0.11/fluent.runtime/fluent/runtime/__init__.py#L140). I'd rather not have to do this.

Are message and term references looked up at runtime or inlined on the compiler branch? I tried to read the code, and failed. I see a dependency check, but I didn't understand the actual generated code.

If they're lookups, compiling on `add_resource` would work? It seems that the lazy compilation isn't doing a lot?

> What the compiler really wants to do is have a constructor that takes all the FTL source in one go, compiles it, then throws away the sources and AST to save memory. It then wants to expose something like the current `format` API. (As an optimization for the really hot inner loops, it might want an API that directly returns the function that a message has been compiled to).

I think that mixing `add_resource` and `get_message` calls is undefined behavior at best, and perfectly fine to discourage. At least in js land, no bundle survives a change to the resource lists, and I think that's good.

In that sense, `add_resource` vs `add_resources` that errors if it's called twice vs arguments to a constructor all sound fair to me. I'd open that conversation in Q4 the earliest, though.

**Comment by Pike** [comment]
The idea of how the different APIs come together is taking shape in https://github.com/Pike/python-fluent/pull/1. I bet that PR gets destroyed once I change it's target back into the projectfluent org, though.

**Comment by spookylukey** [comment]
> > 3. We also need a run-time check for freshness (do we need to re-compile or not?), which for simple messages is a very measurable overhead (about 10%). I've actually eliminated this in my branch, at the cost of some [slightly tricky to follow code](https://github.com/django-ftl/python-fluent/blob/django-ftl-0.11/fluent.runtime/fluent/runtime/__init__.py#L140). I'd rather not have to do this.
> > ```
> 
> Are message and term references looked up at runtime or inlined on the compiler branch? I tried to read the code, and failed. I see a dependency check, but I didn't understand the actual generated code.

Messages are converted into functions, and message references into function calls:

https://github.com/django-ftl/python-fluent/blob/d1481d61e0bc1a28a228a4b6d5258350d436e765/fluent.runtime/tests/test_compiler.py#L93

Terms are always inlined currently, after statically resolving as much as possible:

https://github.com/django-ftl/python-fluent/blob/d1481d61e0bc1a28a228a4b6d5258350d436e765/fluent.runtime/tests/test_compiler.py#L359

(It does this inlining because I have a strong suspicion that this will usually be a memory win, due to the memory overhead of a function call, as well as a time win).

> If they're lookups, compiling on `add_resource` would work? It seems that the lazy compilation isn't doing a lot?

Without lazy compilation, then N calls to `add_resource` will have O(N^2) performance, due to compiling the the first resource N times, the second resource N-1 times etc. We can't (easily) avoid re-compilation of earlier messages, due to inter-resource message/term references that change how something might be compiled.

> I think that mixing `add_resource` and `get_message` calls is undefined behavior at best, and perfectly fine to discourage. At least in js land, no bundle survives a change to the resource lists, and I think that's good.

I'm not sure what you mean - from [here](https://github.com/projectfluent/fluent.js/blob/master/fluent-bundle/src/bundle.js) it looks like multiple calls to `addResource` are allowed?

> 
> In that sense, `add_resource` vs `add_resources` that errors if it's called twice vs arguments to a constructor all sound fair to me. I'd open that conversation in Q4 the earliest, though.

If we have an API that requires you to first instantiate the bundle, then call `add_resources`, and disallows multiple calls to `add_resources`, it's pretty clear to me that the parameters to `add_resources` have to go into the constructor - are we also open to changing the JS API on this front?

There are some other considerations that are also pushing me to thinking that the compiler implementation would be best served by a different API:

* Once I've added support for multiple locales, as I have in django-ftl, for best performance the caching I find I want to do isn't supported by either the old or new `FluentBundle` API. What I really want to do is cache the function generated by the compiler using a key of `(locale, message reference)`, where the reference is `"message"` or `"message.attribute"`. This would require using internals of the compiling `FluentBundle` API at the moment.
* I want a method for returning compilation error messages, ahead of any messages being used (which I've added - it's an addition to the API but a bit awkward at the moment)
* I really want to generate much better error messages using filenames and line numbers of original FTL source files, which isn't possible with the `add_resource` API.

File: fluent.runtime/fluent/runtime/__init__.py [hunk 1]
```diff
@@ -77,12 +86,11 @@ def format(self, message_id, args=None):
             fluent_args = {}
 
         errors = []
-        resolve = self.lookup(message_id)
         env = ResolverEnvironment(context=self,
                                   current=CurrentEnvironment(args=fluent_args),
                                   errors=errors)
         try:
-            result = resolve(env)
+            result = pattern(env)
```
[the remark this answers is not in the document]
Comment (Pike) [reply fluent.runtime/fluent/runtime/__init__.py]: So, `get_message` returns a compiled message, which eventually returns an instance of `resolver.Message`, with the message and attribute values being a subclass of `resolver.BaseResolver`, which all implement the `__call__` method.

**Pull request merged**